qiskit/pulse/ir/pulse_ir.py

```diff
@@ -0,0 +1,400 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=cyclic-import
+
+"""
+=========
+Pulse IR
+=========
+
+"""
+
+from typing import Union, Optional, List, Set, Dict
+
+import numpy as np
+
+from qiskit.pulse.exceptions import PulseError
+
+from qiskit.pulse import (
+    LogicalElement,
+    Frame,
+    MixedFrame,
+    Qubit,
+)
+
+from qiskit.pulse.transforms import AlignmentKind, AlignSequential
+
+from .ir_instructions import (
+    BaseIRInstruction,
+    GenericInstruction,
+    AcquireInstruction,
+)
+
+
+class PulseIR:
+    """
+    ``PulseIR`` is the backbone of the  intermediate representation used in the Qiskit Pulse compiler.
+    A pulse program is represented as a single ``PulseIR`` object, with elements
+    which include IR instructions (base class :class:`~.BaseIRInstruction) and other
+    nested ``PulseIR`` objects. This structure mimics that of :class:`.qiskit.pulse.ScheduleBlock`
+    which is the main pulse program format used in Qiskit.
+
+    The IR is ment to support compilation needs, including partial instructions broadcasting,
+    scheduling, validation and more.
+    """
+
+    def __init__(self, alignment: Optional[AlignmentKind] = AlignSequential):
+        """Create ``PulseIR`` object
+
+        Args:
+            alignment(Optional): The ``AlignmentKind`` to be used for scheduling. Defaults
+                to ``AlignSequential``.
+        """
+        self._elements = []
+        self._alignment = alignment
+
+    @property
+    def initial_time(self) -> int:
+        """Return the initial time ``initial_time`` of the object"""
+        elements_initial_times = [element.initial_time for element in self._elements]
+        if None in elements_initial_times:
+            return None
+        else:
+            return min(elements_initial_times, default=None)
+
+    @property
+    def final_time(self) -> int:
+        """Return the final time of the ``PulseIR``object
+
+        The final time is defined as ``initial_time+duration``"""
+        elements_final_times = [element.final_time for element in self._elements]
+        if None in elements_final_times:
+            return None
+        else:
+            return max(elements_final_times, default=None)
+
+    @property
+    def elements(self) -> List[Union[GenericInstruction, AcquireInstruction, "PulseIR"]]:
+        """Return the elements of the ``PulseIR`` object"""
+        return self._elements
```

Comment:
It seems like there should be a generic type returned that encompasses all of these rather than a list of types.

Reply:
Thanks. I will change that.